File: Linked List/cpp/middleOfList.cpp
// Problem Statement: Given the head of a singly linked list, return the middle node of the linked list. If there are two middle nodes,
// return the second middle node.

// Problem Link: https://leetcode.com/problems/middle-of-the-linked-list/

#include <iostream>
using namespace std;

class ListNode {
public:
    int val;
    ListNode* next;
    ListNode(int val) {
        this->val = val;
        this->next = NULL;
    }
};

int getLength(ListNode* head) {
    int count = 0;
    while (head != NULL) {
        head = head->next;
        count++;
    }
    return count;
}

ListNode* middleNode(ListNode* head) {
    int len = getLength(head);
    
    if (len % 2 != 0) {
        int i = 1;
        while (head != NULL && i <= len / 2) {
            head = head->next;
            i++;
        }
        return head;
    } else {
        int i = 1;
        while (head != NULL && i < len / 2 + 1) {
            head = head->next;
            i++;
        }
        return head;
    }
}

File: Linked List/python/middleOfList.py
# Problem Statement: Given the head of a singly linked list, return the middle node of the linked list. If there are two middle nodes,
# return the second middle node.

# Problem Link: https://leetcode.com/problems/middle-of-the-linked-list/

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def get_length(head):
    count = 0
    while head:
        head = head.next
        count += 1
    return count

def middle_node(head):
    length = get_length(head)
    
    if length % 2 != 0:
        i = 1
        while head and i <= length // 2:
            head = head.next
            i += 1
        return head
    else:
        i = 1
        while head and i < length // 2 + 1:
            head = head.next
            i += 1
        return head
